### PR TITLE
Resolve: [Ledger] Add input for wallet index

### LINKED ISSUE #1786: [Ledger] Add input for wallet index
On Ledger Binance supports different addresses, all can be selected by an `index` (aka `walletIndex`). Add an input field at the right hand side of "ADD LEDGER", so an user can enter a value for an index. Default value is `0`

Screenshot from 2021-09-21 20-33-10

Check Binance.org https://testnet.binance.org/en/unlock or any other wallet as an example.

Screenshot from 2021-09-21 20-50-31

^ This is just for demonstrate how Binance is dealing with it. We do need just an input field - the user can enter any index he want.

### PR DESCRIPTION
**Pull request by gromxyz**
Closes #1786 
Screenshot 2021-09-29 at 11 46 12

### CODE REVIEW

File: src/renderer/services/wallet/ledger.ts [hunk 1]
```diff
@@ -47,19 +57,35 @@ export const createLedgerService = ({ keystore$ }: { keystore$: KeystoreState$ }
   const getLedgerAddress$ = (chain: Chain, network: Network): LedgerAddressLD =>
     FP.pipe(
       ledgerAddresses$,
-      RxOp.map((addressesMap) => addressesMap[chain]),
-      RxOp.distinctUntilChanged(eqLedgerAddressMap.equals),
+      RxOp.map((addressesMap) => addressesMap[chain]['addresses']),
+      // TODO adjust
+      // RxOp.distinctUntilChanged(eqLedgerAddressMap.equals),
```
Comment (gromxyz) [inline src/renderer/services/wallet/ledger.ts]: Update `eqLedgerAddressMap.equals`

File: src/renderer/services/wallet/types.ts [hunk 2]
```diff
@@ -166,5 +168,11 @@ export type LedgerAddressLD = LiveData<LedgerError, Address>
 export type LedgerAddressMap = Record<Network, LedgerAddressRD>
 export type LedgerAddressMap$ = Rx.Observable<LedgerAddressMap>
 
-export type LedgerAddressesMap = Record<Chain, LedgerAddressMap>
+export type LedgerAddressesAndWalletIndexMap = {
+  addresses: LedgerAddressMap
+  walletIndex: string
+}
+export type LedgerAddressesAndWalletIndexMap$ = Rx.Observable<LedgerAddressesAndWalletIndexMap>
```
Comment (gromxyz) [inline src/renderer/services/wallet/types.ts:175]: Better naming probably

File: src/renderer/services/wallet/balances.ts [hunk 3]
```diff
@@ -236,9 +239,10 @@ export const createBalancesService = ({
               // Load balances by given Ledger address
               // and put it's RD state into `balances` of `ChainBalance`
               FP.pipe(
-                getBalanceByAddress$(address, 'ledger'),
-                RxOp.map<WalletBalancesRD, ChainBalance>((balances) => ({
+                Rx.combineLatest([getBalanceByAddress$(address, 'ledger'), getWalletIndex$(chain)]),
+                RxOp.map<any, ChainBalance>(([balances, walletIndex]) => ({
```
Comment (gromxyz) [inline src/renderer/services/wallet/balances.ts]: Don't forget to fix this

File: src/main/api/ledger/address.ts [hunk 4]
```diff
@@ -10,7 +10,8 @@ import { getAddress as getTHORAddress } from './thorchain/address'
 
 export const getAddress = async ({
   chain,
-  network
+  network,
+  walletIndex = '0'
```
Comment (veado) [inline src/main/api/ledger/address.ts]: Use `walletIndex: number` everywhere as we do it in `xchain-*` (e.g. here https://github.com/xchainjs/xchainjs-lib/blob/c17c625b666c0113779db62db2f585c076be587d/packages/xchain-client/src/types.ts#L55)

File: src/renderer/components/wallet/assets/AssetsTableCollapsable.tsx [hunk 5]
```diff
@@ -324,12 +339,24 @@ export const AssetsTableCollapsable: React.FC<Props> = (props): JSX.Element => {
...
-          (_: ApiError) => intl.formatMessage({ id: 'common.error' }),
+          () => {
+            return ''
+          },
+          () => {
+            return intl.formatMessage({ id: 'common.loading' })
+          },
+          (_: ApiError) => {
+            return intl.formatMessage({ id: 'common.error' })
+          },
```
Comment (veado) [inline src/renderer/components/wallet/assets/AssetsTableCollapsable.tsx]: [nitpick] Changes to use `return` are not needed

File: src/renderer/components/wallet/settings/WalletSettings.tsx [hunk 6]
```diff
@@ -96,11 +97,21 @@ export const WalletSettings: React.FC<Props> = (props): JSX.Element => {
...
-          <Styled.AddLedgerIcon /> {intl.formatMessage({ id: 'ledger.add.device' })}
-        </Styled.AddLedgerButton>
+        <Styled.AddLedgerContainer>
+          <Styled.AddLedgerButton loading={loading} onClick={() => addLedgerAddress(chain, walletIndex)}>
+            <Styled.AddLedgerIcon /> {intl.formatMessage({ id: 'ledger.add.device' })}
+          </Styled.AddLedgerButton>
+          {isBnbChain(chain) && (
+            <Styled.WalletIndexInput
+              value={walletIndex.toString()}
+              pattern="[0-9.]+"
```
Comment (veado) [inline src/renderer/components/wallet/settings/WalletSettings.tsx]: Remove `.` Maybe regex `\d` (digits only) will work, too?

File: src/renderer/services/wallet/ledger.ts [hunk 7]
```diff
@@ -47,19 +58,26 @@ export const createLedgerService = ({ keystore$ }: { keystore$: KeystoreState$ }
...
-      RxOp.map((addressesMap) => addressesMap[chain]),
+      RxOp.map((addressesMap) => addressesMap[chain]['addresses']),
       RxOp.distinctUntilChanged(eqLedgerAddressMap.equals),
       RxOp.map((addressMap) => addressMap[network])
     )
 
+  const getWalletIndex$ = (chain: Chain): Rx.Observable<string> =>
+    FP.pipe(
+      ledgerAddresses$,
+      RxOp.map((addressesMap) => addressesMap[chain]['walletIndex'])
```
Comment (veado) [inline src/renderer/services/wallet/ledger.ts]: `addressesMap[chain].walletIndex`

File: src/renderer/services/wallet/ledger.ts [hunk 8]
```diff
@@ -47,19 +58,26 @@ export const createLedgerService = ({ keystore$ }: { keystore$: KeystoreState$ }
   const getLedgerAddress$ = (chain: Chain, network: Network): LedgerAddressLD =>
     FP.pipe(
       ledgerAddresses$,
-      RxOp.map((addressesMap) => addressesMap[chain]),
+      RxOp.map((addressesMap) => addressesMap[chain]['addresses']),
```
Comment (veado) [inline src/renderer/services/wallet/ledger.ts]: `addressesMap[chain].addresses`

**Review by veado — CHANGES_REQUESTED** [review]

File: src/renderer/components/wallet/settings/WalletSettings.tsx (hunk 6 above)
Comment (gromxyz) [reply src/renderer/components/wallet/settings/WalletSettings.tsx]: This one is maybe a bit more readable?

**Review by veado — DISMISSED** [review]
:+1:

**Pull request merged**